### ISSUE TITLE
different jpg image get different condition

### ISSUE THREAD

**Issue by zhedahe** [moured/YOLO-ONNX-Parse-Example#1]
hi,  as for inputting into **zidane.jpg**, I found pth result by pth inference is not same with onnxruntime inference result, **but bus.jpg(your sample) is same**. pls help me to valid it, thanks!

**Comment by moured** [comment]
hello @zhedahe, could you please attach the image "zidane.jpg" and what you get as output

**Comment by zhedahe** [comment]
> 
> 
> hello @zhedahe, could you please attach the image "zidane.jpg" and what you get as output

ok, I met 2 problems: 
1) right person detection in zidane.jpg wiill **miss a tie** by using your code, zidane.jpg and its result can be seen in attachment（zidane.jpg is origin image, zidane_reisze.jpg is detection image using you code,  zidane_allsize.jpg is no-resized original detection image）. 
2) I try to scale coords from resize img to origin img, but positions of detection **are not accurat**e, **there are some deviations**(pls check zidane_allsize.jpg).
here is my all complete code, first part is pth inference, it is all right by refering yolov5  code from **ultralytics/yolov5**. second part is onnx inference by your code, I mad some change, in the end of second part is scaling coords from resize img to full-scale img
thank for your help!
![zidane](https://user-images.githubusercontent.com/4446863/117781164-eb61c080-b272-11eb-9a76-7c1ae4064d5e.jpg)
![zidane_allsize](https://user-images.githubusercontent.com/4446863/117781172-ed2b8400-b272-11eb-9144-624e8d83ece8.jpg)
![zidane_resize](https://user-images.githubusercontent.com/4446863/117781175-edc41a80-b272-11eb-9bad-e955fc6d7a67.jpg)

#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import time
import glob
from PIL import Image, ImageDraw
import numpy as np
import cv2
import math 
import csv
import time 
import glob 
import matplotlib.pyplot as plt

import onnx 
import onnxruntime
import torch
import torch.nn as nn
import torch.nn.functional as F
from torchvision import transforms, models
from utils.torch_utils import select_device, time_synchronized
from models.experimental import attempt_load
from utils.datasets import LoadImages
from utils.general import non_max_suppression, check_img_size, scale_coords
from utils.plots import colors, plot_one_box
from utils.datasets import letterbox


if __name__ == "__main__":
    source = 'data/images' # images dir
    weights = 'yolov5s.pt'
    imgsz = 640
    device = select_device('0')
    half = device.type != 'cpu'  # half precision only supported on CUDA
    
    # Load model
    print('loading pth model...........')
    t1 = time_synchronized()
    model = attempt_load(weights, map_location=device)  # load FP32 model
    t2 = time_synchronized()
    print ('load model time: {} ms'.format(1000*(t2-t1)))
    stride = int(model.stride.max())  # model stride
    imgsz = check_img_size(imgsz, s=stride)  # check img_size
    names = model.module.names if hasattr(model, 'module') else model.names  # get class names
    print('img size:', imgsz)
    print('stride range:', model.stride)
    # print ('names:', names)
    if half:
        model.half()  # to FP16

    # Run inference
    if device.type != 'cpu':
        model(torch.zeros(1, 3, imgsz, imgsz).to(device).type_as(next(model.parameters())))  # run once

    for img_path in glob.glob('{}/*.jpg'.format(source)):
        im0s = cv2.imread(img_path)
        img = letterbox(im0s, imgsz, stride)[0] 

        img = img[:, :, ::-1].transpose(2, 0, 1)  # BGR to RGB
        print ('make border img shape:', img.shape)
        img = np.ascontiguousarray(img)

        img = torch.from_numpy(img).to(device)
        img = img.half() if half else img.float()  # uint8 to fp16/32
        img /= 255.0  # 0 - 255 to 0.0 - 1.0
        if img.ndimension() == 3:
            img = img.unsqueeze(0)

        # Inference
        t1 = time_synchronized()
        pred = model(img, augment=False)[0]

        # Apply NMS
        pred = non_max_suppression(pred, 0.25, 0.45, classes=None, agnostic=False)
        t2 = time_synchronized()
        print ('inference time: {} ms'.format(1000*(t2-t1)))
        # print ('res by pth:', pred)
        
        xyxy = []
        # Process detections
        for i, det in enumerate(pred):
            im0 = im0s.copy()
            # Rescale boxes from img_size to im0 size
            det[:, :4] = scale_coords(img.shape[2:], det[:, :4], im0.shape).round()

            for *xyxy, conf, cls in reversed(det):
                c = int(cls)  # integer class
                # label = '{} {}'.format(names[c], conf)
                label = ''
            
                img_name = img_path.split('/')[-1]
                plot_one_box(xyxy, im0, label=label, color=colors(c, True), line_thickness=3)
                cv2.imwrite('runs/pth_imgs/{}'.format(img_name), im0)

    print('---- pth over------\n')
    # check onnx model
    onnx_model = onnx.load('yolov5s.onnx')
    try:
        onnx.checker.check_model(onnx_model)
    except (Exception, err):
        print('model error:', err)
    else:
        print('model is ok!')
    finally:
        print('modle chekc is over!')

    session = onnxruntime.InferenceSession("yolov5s.onnx")
    batch_size = session.get_inputs()[0].shape[0]
    img_size_h = session.get_inputs()[0].shape[2]
    img_size_w = session.get_inputs()[0].shape[3]
    print ('batch_size: {}, h:{}, w:{}'.format(batch_size, img_size_h ,img_size_w))

    for image_name in os.listdir(source):
        image_path = os.path.join(source, image_name)
        cv2_img = cv2.imread(image_path)
        
        """
        problem one:
        follow pth's operation: using padding image as line.61 code
        if comment this line.123 code, right persion in zidane.jpg  will miss a tie detection 
        """
        cv2_img = letterbox(cv2_img, imgsz, stride)[0] 


        cv2_img = cv2.cvtColor(cv2_img, cv2.COLOR_BGR2RGB)
        cv2_img = np.ascontiguousarray(cv2_img)
        image_src = Image.fromarray(cv2_img)
        print('Loaded Image Info : ',image_src.format, image_src.size, image_src.mode)

        resized =  image_src.resize((img_size_w, img_size_h))
        print('After resizing :' ,resized.size, resized.mode)

        img_in = np.transpose(resized, (2, 0, 1)).astype(np.float32) # HWC -> CHW
        img_in = np.expand_dims(img_in, axis=0)
        img_in /= 255.0
        print('Batch-Size, Channel, Height, Width : ',img_in.shape)

        input_name = session.get_inputs()[0].name
        t1 = time_synchronized()
        outputs = session.run(None, {input_name: img_in})

        print('Concatenated outputs shape: ', outputs[0].shape)
        print('Separated outputs shape: ', outputs[1].shape, outputs[2].shape, outputs[3].shape)

        filterd_predictions = non_max_suppression(torch.tensor(outputs[0]), conf_thres = 0.25, iou_thres = 0.45)
        t2 = time_synchronized()
        print ('inference time: {} ms'.format(1000*(t2-t1)))

        # draw detection results in resize image
        img = cv2.resize(cv2.imread(image_path), (img_size_w, img_size_h), interpolation=cv2.INTER_LINEAR)
        for pred in filterd_predictions[0]:
            pred = pred.detach().cpu().numpy()
            x1 = int(pred[0])
            y1 = int(pred[1])
            x2 = int(pred[2])
            y2 = int(pred[3])
            img = cv2.rectangle(img, (x1,y1),(x2,y2), thickness = 2, color=(0,255,0))
            print('Confidence: ', pred[4], 'Class:', int(pred[5]))
        img_name = image_path.split('/')[-1]
        img_name = img_name.replace('.jpg', '_resize.jpg')
        cv2.imwrite('runs/ox_imgs/{}'.format(img_name), img)

        # draw detection results in original image
        """
        Poblem two:
        after scaling coords from resize img to original img, I found positions of all detetion targets are not accurate
        """
        xyxy = []
        for i, det in enumerate(filterd_predictions):
            im0 = cv2.imread(image_path)
            det[:, :4] = scale_coords([img_size_h, img_size_w], det[:, :4], im0.shape).round()

            for *xyxy, conf, cls in reversed(det):
                c = int(cls)  # integer class
                # label = '{} {}'.format(names[c], conf)
                label = ''
                img_name = image_path.split('/')[-1]
                plot_one_box(xyxy, im0, label=label, color=colors(c, True), line_thickness=3)
                cv2.imwrite('runs/ox_imgs/{}'.format(img_name), im0)

    print('---- onnx over------\n')

**Comment by zhedahe** [comment]
ok, I met 2 problems:
    1. right person detection in zidane.jpg wiill **miss a tie** by using your code, zidane.jpg and its result can be seen in attachment（zidane.jpg is origin image, zidane_reisze.jpg is detection image using you code, zidane_allsize.jpg is **no-resized** original detection image）.
    2. I try to scale coords from resize img to origin img, but positions of detection are **not accurate, there are some deviations**(pls check zidane_allsize.jpg).
    here is my all complete code, first part is **pth inference**, it is all right by refering yolov5 code from ultralytics/yolov5. second part is **onnx inference** by your code, I mad some change, in the end of second part is scaling coords from resize img to full-scale img
    pls help me to check above problem, I will try to find methods at the same. thank your very much!

zidane  predict result
![zidane](https://user-images.githubusercontent.com/4446863/117782143-ec472200-b273-11eb-915d-b7436868d081.jpg)

zidane_allsize predict result 
![zidane_allsize](https://user-images.githubusercontent.com/4446863/117782146-ed784f00-b273-11eb-8318-ee4961bb233c.jpg)

zidane_resize predict result
![zidane_resize](https://user-images.githubusercontent.com/4446863/117782152-ee10e580-b273-11eb-954b-2efb66ed3f4c.jpg)

**Comment by zhedahe** [comment]
#!/usr/bin/env python3
    #-*- coding: utf-8 -*-
    
    import os
    import time
    import glob
    from PIL import Image, ImageDraw
    import numpy as np
    import cv2
    import math 
    import csv
    import time 
    import glob 
    import matplotlib.pyplot as plt
    
    import onnx 
    import onnxruntime
    import torch
    import torch.nn as nn
    import torch.nn.functional as F
    from torchvision import transforms, models
    from utils.torch_utils import select_device, time_synchronized
    from models.experimental import attempt_load
    from utils.datasets import LoadImages
    from utils.general import non_max_suppression, check_img_size, scale_coords
    from utils.plots import colors, plot_one_box
    from utils.datasets import letterbox
    
    
    if __name__ == "__main__":
         source = 'data/images' # images dir
         weights = 'yolov5s.pt'
         imgsz = 640
         device = select_device('0')
         half = device.type != 'cpu'  # half precision only supported on CUDA
        
        # Load model
        print('loading pth model...........')
        t1 = time_synchronized()
        model = attempt_load(weights, map_location=device)  # load FP32 model
        t2 = time_synchronized()
        print ('load model time: {} ms'.format(1000*(t2-t1)))
        stride = int(model.stride.max())  # model stride
        imgsz = check_img_size(imgsz, s=stride)  # check img_size
        names = model.module.names if hasattr(model, 'module') else model.names  # get class names
        print('img size:', imgsz)
        print('stride range:', model.stride)
        # print ('names:', names)
        if half:
            model.half()  # to FP16
    
        # Run inference
        if device.type != 'cpu':
            model(torch.zeros(1, 3, imgsz, imgsz).to(device).type_as(next(model.parameters())))  # run once
    
        for img_path in glob.glob('{}/*.jpg'.format(source)):
            im0s = cv2.imread(img_path)
            img = letterbox(im0s, imgsz, stride)[0] 
    
            img = img[:, :, ::-1].transpose(2, 0, 1)  # BGR to RGB
            print ('make border img shape:', img.shape)
            img = np.ascontiguousarray(img)
    
            img = torch.from_numpy(img).to(device)
            img = img.half() if half else img.float()  # uint8 to fp16/32
            img /= 255.0  # 0 - 255 to 0.0 - 1.0
            if img.ndimension() == 3:
                img = img.unsqueeze(0)
    
            # Inference
            t1 = time_synchronized()
            pred = model(img, augment=False)[0]
    
            # Apply NMS
            pred = non_max_suppression(pred, 0.25, 0.45, classes=None, agnostic=False)
            t2 = time_synchronized()
            print ('inference time: {} ms'.format(1000*(t2-t1)))
            # print ('res by pth:', pred)
            
            xyxy = []
            # Process detections
            for i, det in enumerate(pred):
                im0 = im0s.copy()
                # Rescale boxes from img_size to im0 size
                det[:, :4] = scale_coords(img.shape[2:], det[:, :4], im0.shape).round()
    
                for *xyxy, conf, cls in reversed(det):
                    c = int(cls)  # integer class
                    # label = '{} {}'.format(names[c], conf)
                    label = ''
                
                    img_name = img_path.split('/')[-1]
                    plot_one_box(xyxy, im0, label=label, color=colors(c, True), line_thickness=3)
                    cv2.imwrite('runs/pth_imgs/{}'.format(img_name), im0)
    
        print('---- pth over------\n')
        # check onnx model
        onnx_model = onnx.load('yolov5s.onnx')
        try:
            onnx.checker.check_model(onnx_model)
        except (Exception, err):
            print('model error:', err)
        else:
            print('model is ok!')
        finally:
            print('modle chekc is over!')
    
        session = onnxruntime.InferenceSession("yolov5s.onnx")
        batch_size = session.get_inputs()[0].shape[0]
        img_size_h = session.get_inputs()[0].shape[2]
        img_size_w = session.get_inputs()[0].shape[3]
        print ('batch_size: {}, h:{}, w:{}'.format(batch_size, img_size_h ,img_size_w))
    
        for image_name in os.listdir(source):
            image_path = os.path.join(source, image_name)
            cv2_img = cv2.imread(image_path)
            
            """
            problem one:
            follow pth's operation: using padding image as line.61 code
            if comment this line.123 code, right persion in zidane.jpg  will miss a tie detection 
            """
            # cv2_img = letterbox(cv2_img, imgsz, stride)[0] 
    
    
            cv2_img = cv2.cvtColor(cv2_img, cv2.COLOR_BGR2RGB)
            cv2_img = np.ascontiguousarray(cv2_img)
            image_src = Image.fromarray(cv2_img)
            print('Loaded Image Info : ',image_src.format, image_src.size, image_src.mode)
    
            resized =  image_src.resize((img_size_w, img_size_h))
            print('After resizing :' ,resized.size, resized.mode)
    
            img_in = np.transpose(resized, (2, 0, 1)).astype(np.float32) # HWC -> CHW
            img_in = np.expand_dims(img_in, axis=0)
            img_in /= 255.0
            print('Batch-Size, Channel, Height, Width : ',img_in.shape)
    
            input_name = session.get_inputs()[0].name
            t1 = time_synchronized()
            outputs = session.run(None, {input_name: img_in})
    
            print('Concatenated outputs shape: ', outputs[0].shape)
            print('Separated outputs shape: ', outputs[1].shape, outputs[2].shape, outputs[3].shape)
    
            filterd_predictions = non_max_suppression(torch.tensor(outputs[0]), conf_thres = 0.25, iou_thres = 0.45)
            t2 = time_synchronized()
            print ('inference time: {} ms'.format(1000*(t2-t1)))
    
            # draw detection results in resize image
            img = cv2.resize(cv2.imread(image_path), (img_size_w, img_size_h), interpolation=cv2.INTER_LINEAR)
            for pred in filterd_predictions[0]:
                pred = pred.detach().cpu().numpy()
                x1 = int(pred[0])
                y1 = int(pred[1])
                x2 = int(pred[2])
                y2 = int(pred[3])
                img = cv2.rectangle(img, (x1,y1),(x2,y2), thickness = 2, color=(0,255,0))
                print('Confidence: ', pred[4], 'Class:', int(pred[5]))
            img_name = image_path.split('/')[-1]
            img_name = img_name.replace('.jpg', '_resize.jpg')
            cv2.imwrite('runs/ox_imgs/{}'.format(img_name), img)
    
            # draw detection results in original image
            """
            Poblem two:
            after scaling coords from resize img to original img, I found positions of all detetion targets are not accurate
            """
            xyxy = []
            for i, det in enumerate(filterd_predictions):
                im0 = cv2.imread(image_path)
                det[:, :4] = scale_coords([img_size_h, img_size_w], det[:, :4], im0.shape).round()
    
                for *xyxy, conf, cls in reversed(det):
                    c = int(cls)  # integer class
                    # label = '{} {}'.format(names[c], conf)
                    label = ''
                    img_name = image_path.split('/')[-1]
                    plot_one_box(xyxy, im0, label=label, color=colors(c, True), line_thickness=3)
                    cv2.imwrite('runs/ox_imgs/{}'.format(img_name), im0)
    
        print('---- onnx over------\n')

**Comment by zhedahe** [comment]
@moured  some information is listed above, pls help me to validate it, thanks a lot!

**Comment by moured** [comment]
> @moured some information is listed above, pls help me to validate it, thanks a lot!

hello @zhedahe I am working on it right now. Will update you very soon

**Comment by moured** [comment]
> ok, I met 2 problems:
> 1. right person detection in zidane.jpg wiill **miss a tie** by using your code, zidane.jpg and its result can be seen in attachment（zidane.jpg is origin image, zidane_reisze.jpg is detection image using you code, zidane_allsize.jpg is **no-resized** original detection image）.
> 2. I try to scale coords from resize img to origin img, but positions of detection are **not accurate, there are some deviations**(pls check zidane_allsize.jpg).
> here is my all complete code, first part is **pth inference**, it is all right by refering yolov5 code from ultralytics/yolov5. second part is **onnx inference** by your code, I mad some change, in the end of second part is scaling coords from resize img to full-scale img
> pls help me to check above problem, I will try to find methods at the same. thank your very much!
> 
> zidane predict result
> ![zidane](https://user-images.githubusercontent.com/4446863/117782143-ec472200-b273-11eb-915d-b7436868d081.jpg)
> 
> zidane_allsize predict result
> ![zidane_allsize](https://user-images.githubusercontent.com/4446863/117782146-ed784f00-b273-11eb-8318-ee4961bb233c.jpg)
> 
> zidane_resize predict result
> ![zidane_resize](https://user-images.githubusercontent.com/4446863/117782152-ee10e580-b273-11eb-954b-2efb66ed3f4c.jpg)

Thanks, @zhedahe for sharing the detailed steps.

For problem 1 : 
unfortunately, the original network is trained with their padding approach. However, if you reduce the confidence threshold in the nmx function to 0.1 or higher a bit you will get the tie [TESTED]. If you don't want to use a lower conf threshold in your project then you need to retrain the network with commenting padding part. 

For problem 2:
please try this solution it works: https://stackoverflow.com/questions/49466033/resizing-image-and-its-bounding-box

**Comment by moured** [comment]
Let me know if any updates you have. Thank you

**Comment by zhedahe** [comment]
> Let me know if any updates you have. Thank you
thanks for your help, I had solved 2 problems by your advice, thank you!
